feat: BEMIT documentation (#176)
sparkbox/standard: code-style/scss/README.md

# SCSS

**Table of Contents**

1. [Introduction](#Introduction)
2. [Class Naming Theory](#Class-Naming-Theory)
3. [BEM](#BEM)
  1. [Using BEM](#Using-BEM)

>     1. [Block](#Block)

Reviewer: ✊ GITHUB!!!

    2. [Element](#Element)
    3. [Modifier](#Modifier)
  2. [BEM in Sass](#BEM-in-Sass)
4. [ITCSS](#ITCSS)
  1. [File Naming](#File-Naming)
    1. [Settings](#Settings)
    2. [Tools](#Tools)
    3. [Generic](#Generic)
    4. [Elements](#Elements)
    5. [Objects](#Objects)
    6. [Components](#Components)
    7. [Utilities](#Utilities)
5. [Name Spacing](#Name-Spacing)
6. [ITCSS and BEM](#ITCSS-and-BEM)

## Introduction

> Our CSS is organized using [Harry Roberts](csswizardry.com)' Inverted Triangle CSS (ITCSS) organizational approach. This method is mixed with [Block Element Modifier](http://getbem.com) (BEM) naming convention for class names throughout the stylesheet.

Reviewer 2: I think we prefer the `[Link Text]` ... `[Link Text]: link` syntax for links so they are more readable.

Reviewer 2: the last one here
https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#links

> 

Reviewer 2: I really think we should also link https://seesparkbox.com/foundry/bem_by_example somewhere in this page. Mayeb a "read more" section is in order? 

also https://seesparkbox.com/foundry/thoughtful_css_architecture

Reviewer: Definitely. Good point. It crossed my mind, but I kept for getting to work it in.

---

> ## Class Naming Theory

Reviewer 2: LOVE that you added this section

When it comes to naming classes, we find it best to be as specific as possible with the fewest words possible. This can be an exercise in understanding the purpose and function of a component. It's helpful to find like components and think about what they have in common and what their purpose is. You can then choose a generic name that encapsulates the uses the component will have.

Keep in mind that naming is for humans, so verboseness is useful if it helps communicate what the class does. Ideally, a future developer should be able to look at a class name and have a reasonable idea of what they might see on the rendered page.

> 

Reviewer 2: I would love an example here! 

What is a good use case for using a full word in a class name? What is a good example of abbreviation?

---

## BEM

> BEM stands for block element modifier, which is the order you follow for naming with BEM.

Reviewer 2: ```suggestion
BEM stands for "block element modifier", which is the order you follow for naming with BEM.
```

#### Block

> `.block` is the parent wrapper of the component. This would be something like site navigation, the hero section, or a footer. Any container that holds other code could be a block.

Reviewer 2: I'd love an example!

Reviewer 2: Is there a case for naming a block that doesn't hold other code?

#### Element

> `.block__element` is an element inside the component, or inside the parent wrapper. This could be a title or a link. It's anything inside the parent wrapper. It's important here to note that BEM isn't meant to reflect nesting. For example, say you have a section that includes an image element inside an anchor element:

Reviewer 2: Maybe mention that `__` is two underscores, it's a little hard to read in the github page text.

```html
<section>
  <a href="#">
    <img src="#">
  </a>

> </section>

Reviewer 2: Wonderful example!

```

Our BEM naming doesn't need to show that we nested that image tag, so we would give it a class name of something like `.block__image`. We would _not_ give it a class name of `.block__link__image`.

#### Modifier

> Modifiers are the most specific of the three parts of BEM. They should only define one or two properties that change the previously defined properties on the block or element. You can use a modifier on a block or an element.

Reviewer 2: The way this paragraph starts is inconsistant with the other two. Though, I like the syntax here, where you use the name "Modifier" as a word and show the syntax later on - it's less confusing.

`.block--modifier` modifies the entire component.

`.block__element--modifier` modifies a single element inside a component.

> #### Example

Reviewer 2: Ahh, here we go.

> Let's look at an example. Maybe we have a hero section that's reused on each page of a site. The block might be called `.hero`, because that's what the entire section is. The title, description, and button are all elements, so they would be `.hero__title`, `.hero__description`, and `.hero__button`. We don't need any modifiers yet, because a modifier is really specific.

Reviewer 2: >  because a modifier is really specific.

is a little arbitrary. Maybe patronizing. How about: 

>  because we have nothing to modify yet.

```html
<section class="hero">
  <h1 class="hero__title"></h1>
  <p class="hero__description"></p>
  <button class="hero__button"></p>
</section>
```

Now, what if on one page where we want a dark version of our button? We can create a modifier class for the `background-color` property so that we're only changing the color of the button with that modifier class. Notice that you still need to include the original `.hero__button` element level class. That's because the modifier is only changing the `background-color`, and you still want all the other properties from the original button class.

```html
<section class="hero">
  <h1 class="hero__title"></h1>
  <p class="hero__description"></p>
  <button class="hero__button hero__button--dark"></p>

> </section>

Reviewer 2: Here is where a section like this could be useful. You have plenty of example of intro BEM, and if anyone is having trouble wrapping their head around the concept, pointing them to the foundry could be a great path forward: 

`Want more examples of BEM? [Check out Nathan's Foundry article, BEM by Example][]`

```

### BEM in Sass

> The formatting of BEM relies on Sass local variables to ease the amount of repetitiveness inherent with BEM. Using the same structure above looks like this in the SCSS file:

Reviewer 2: Would a link to https://sass-lang.com/guide be too basic here?

> The formatting of BEM relies on Sass local variables to ease the amount of repetitiveness inherent with BEM. Using the same structure above looks like this in the SCSS file:

Reviewer 2: Is there a way to phrase this more positively? 

> The formatting of BEM relies on Sass local variables to ease the amount of repetitiveness inherent with BEM.

Maybe 

> In combination with Sass local variables and nested syntax, BEM is an efficient... blah blah

Author: Good call 👍

```scss
.block {
  // styles for .block

  &__element {
    // styles for .block__element

    &--modifier {
      // styles for .block__element--modifier
    }
  }

  &--modifier {
    // styles for .block--modifier
  }

> }

Reviewer 2: Perfect example!

```

Let's use our example from above to see what our Sass would look like:

> <section class="hero">

Reviewer 2: wrap this in ` ```html ` maybe

  <h1 class="hero__title"></h1>
  <p class="hero__description"></p>
  <button class="hero__button hero__button--dark"></p>
</section>

```scss
.hero {
  // styles for .block

  &__title {
    // styles for .block__element
  }

  &__description {
    // styles for .block__element
  }

  &__button {
    // styles for .block__element

    &--dark {
      // styles for .block__element--modifier
    }
  }
}
```

---

## ITCSS

> ITCSS stands for Inverted Triangle CSS. It separates CSS into sections to help developers manage CSS specificity and the cascade. There are seven layers of the inverted triangle. The idea is that you start with a wide reach and low specificity at the top, and as you go down, the reach narrows and the properties get more specific. This helps you avoid specificity issues.

Reviewer 2: Here is where you could link the CSS Architecture article?

> ITCSS stands for Inverted Triangle CSS. It separates CSS into sections to help developers manage CSS specificity and the cascade. There are seven layers of the inverted triangle. The idea is that you start with a wide reach and low specificity at the top, and as you go down, the reach narrows and the properties get more specific. This helps you avoid specificity issues.

Reviewer 2: > There are seven layers of the inverted triangle, **though not every project will need all seven**.

?

#### Settings

When using a preprocessor, settings are used to define variables. This puts variables at the top of the cascade so they are accessible to all partials further down. No CSS selectors should be generated by the preprocessor from partials in this section.

When not using a preprocessor, the settings area is a good place to write comments describing the colors and font stacks as a guide. If using features such as custom properties, those that are global custom properties should be described in this section.

#### Tools

> This section is specifically for preprocessor global mixins and functions. No CSS should be generated by the preprocessor from partials in this section. Some of our favorite global mixins to use are [unbuttonize](https://gist.github.com/zastrow/bb43d2b256d52738e0ffdee7f825e79f) to remove default button styles and [delist](https://gist.github.com/zastrow/94d97dfe2e7e2808bcf4dc45268e6338) to remove default list styles from `<ol>` and `<ul>`.

Reviewer 2: Wonderful examples!

#### Generic

Here we define any generic styles that are not specific to the styles of the site. This section should include any Reset or Normalize libraries and any preferential base styles for elements. There shouldn’t be any classes or ids used in this section.

#### Elements

> This is where we define styling for bare HTML elements (like H1, A, etc.). These elements come with default styling from the browser, so we can redefine them here.

Reviewer 2: ```suggestion
This is where we define styling for bare HTML elements (`h1`, `a`, etc.). These elements come with default styling from the browser, so we can redefine them here.
```

#### Objects

> Objects are class-based selectors that define undecorated design patterns, for example [the media object](http://www.stubbornella.org/content/2010/06/25/the-media-object-saves-hundreds-of-lines-of-code/) (a pattern with an image on the left and descriptive content on the right, like a Tweet for example) from OOCSS.

Reviewer 2: Woooop interesting link

Reviewer 2: What's an undecorated design pattern?

#### Components

> This section is where we define our specific UI components. This is where the majority of our work takes place, and our UI components are often composed of Objects and Components.

Reviewer 2: Example? Something like "A component could be a button or a textarea."

#### Utilities

> Utilities are helper classes with the ability to override anything that goes before it in the triangle. Utility classes are very specific. Sometimes we use utility classes to set up spacing systems, which would consist of creating classes for certain sizes of padding or margin to be used anywhere on the site, or to establish a hide helper class that our JavaScript can interact with to toggle the visibility of elements.

Reviewer 2: > Utilities, sometimes called "trumps",

Reviewer 2: LOVE the example.

Reviewer 2: Maybe note that ... "because of the overriding nature of utility classes, we sometimes choose to prefix the class name with `u-` for larger projects" 

That english is bad but I hope you understand.

Author: Haha, yes, makes sense

---

## Namespacing

> We namespace to articulate what class names we wrote. As with the [class naming section above](#Class-Naming-Theory), namespacing is a useful tool to describe what kind of class this is and how it functions. Also, like class naming, the more information in the namespace, the easier it is to comprehend for future developers.

Reviewer 2: > We namespace to articulate what class names we wrote.

I don't really follow this sentence. I think you could start with the second sentence.

> We namespace to articulate what class names we wrote. As with the [class naming section above](#Class-Naming-Theory), namespacing is a useful tool to describe what kind of class this is and how it functions. Also, like class naming, the more information in the namespace, the easier it is to comprehend for future developers.

Reviewer 2: ```suggestion
As with the [class naming section above](#Class-Naming-Theory), "namespacing" is a useful tool to describe how a class is meant to function. A namespace prevents collisions between classes written by legacy developers, design systems, and/or vendor stylesheets. The more information in the namespace, the easier it is to comprehend for future developers.
```

A namespace title should consist of more than one character but no more than four characters. One character rarely provides enough information to describe the type of class, and more than four characters cause the namespace to compete with the class name. Namespaces should be quick to comprehend but only act as an aid to define the type of styles that will exist and how they function.

Our namespacing is derived from the ITCSS section the class belongs in. Since settings, tools, generic, and elements are only used in file naming (and usually use the whole title in the file name), we have only defined ITCSS prefixes for the sections used for class naming:

- `.obj-` (objects)
- `.cmp-` (components)

> - `.util-` (utilities)

Reviewer 2: I've never used this. Is it really universal?

Reviewer 2: I'd probably actively argue against using this in some of the larger projects we work on. 😬

Also, we often namespace utilities as `u-` or colors as `c-`. Design systems may prefix all classes and this wouldn't work there - for example, a sparkbox design system could be `sbds-` or `sb-`.

Author: We are making an effort to make this universal. The argument I've heard for it is that, like you pointed out, `c-` could mean component or color or anything else. Including a couple extra characters helps to clarify what it is without needing a key or a cheatsheet. @zastrow may have something more to add here.

> It's important to note that these prefixes are only used in _class_ naming. Sometimes we want to prefix a _file_ name with the ITCSS section to communicate what the file includes, and in that case, we would use the entire name of the section.

Reviewer 2: > For example, a file of color variables could be named `._settings.colors`.

---

## How ITCSS works with BEM

> We use ITCSS and BEM together to help us communicate the purpose of a class. ITCSS shows up both in file naming and in class naming, while BEM is only used for class naming.

Reviewer 2: ```suggestion
We often use ITCSS and BEM together to help us communicate the purpose of a class. ITCSS shows up both in file naming and in class naming, while BEM is only used for class naming.
```

I'd maybe avoid using syntax that says there's only _one_ way to do something.

Author: Really good point, thank you 👍

### ITCSS File Naming

> The parts of ITCSS that are used for file naming don't use BEM. An example of this might be color variables. Those belong in the settings section of the inverted triangle, so we would create a file for that called `settings.variables.color.scss` in which we might define some colors:

Reviewer 2: love this section.

```scss
$c-sky-blue: //value for blue

> $c-crimson: //value for red

Reviewer 2: This goes counter to what you mentioned above, about only using namespaces with 2 or more characters.

Reviewer: It does. For consistency, I’m good with using the same convention in variable naming.

```

> We don't need BEM for this because we're not defining classes.

Reviewer 2: But I still do 😁

### ITCSS and BEM Class Naming

Objects, components, and utilities are the parts of ITCSS that are used in class naming, and they work together with BEM. Let's look at our example from the [BEM](#BEM) section again:

```html
<section class="hero">
  <h1 class="hero__title">...</h1>
  <p class="hero__description">...</p>
  <button class="hero__button hero__button--dark">...</button>
</section>
```

Right now, we're only using BEM here, but we can add ITCSS to better communicate to other developers what this piece of code is (a component), and what it's for (defining classes for the hero block). This example is a hero section, which is a specific part of our site's UI. That means it's a component in ITCSS. For components, we use `cmp-` to prefix class names and denote their purpose.

```html
<section class="cmp-hero">
  <h1 class="cmp-hero__title">...</h1>
  <p class="cmp-hero__description">...</p>
  <button class="cmp-hero__button cmp-hero__button--dark">...</button>

> </section>

Reviewer 2: I'm not a huge fan of this section. BUT if we document it as _an_ option not _the only_ option I'm more in favor.

Reviewer 2: If something is a BEM block, it's going to be a component or an object. So why would I need to also say it's a component? It's like saying `block-block__element`

Reviewer 2: Also, most of our projects don't really use objects. There might be layout stuff in objects, but not much more than that.

Reviewer: This has become pretty common from what I have seen. The argument that was made and why I am putting this in here is that the name spacing a component, `cmp-`, is a way to show that is a component *we* made. We will explain that further in the documentation.

```

Maybe we want this component to have some layout styles, like centering it on the page and giving it a max-width. That would require a layout class, which is on the object level of ITCSS. To do this, we can create a stylesheet file called `objects.layout.scss` (remembering that we use the full name of the ITCSS section in file naming and the prefix of the ITCSS section in class naming) and include something like:

```scss
.obj-layout {

>   max-width: 75rem;

Reviewer 2: YES rem
